src/Grpc.Net.Client/Balancer/Internal/ActiveSubchannelTransport.cs

```diff
@@ -0,0 +1,294 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Net;
+using System.Net.Http;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Microsoft.Extensions.Logging;
+
+namespace Grpc.Net.Client.Balancer.Internal
+{
+#if NET5_0_OR_GREATER
+    internal class ActiveSubchannelTransport : ISubchannelTransport, IDisposable
+    {
+        private readonly SemaphoreSlim _connectionCreateLock;
+        private readonly Subchannel _subchannel;
+        private readonly TimeSpan _socketPingInterval;
+        private int _lastEndPointIndex;
+
+        internal readonly List<(DnsEndPoint EndPoint, Socket Socket, Stream? Stream)> _activeStreams;
+        private readonly Timer _socketConnectedTimer;
+        private Socket? _initialSocket;
+        private DnsEndPoint? _initialSocketEndPoint;
+        private bool _disposed;
+        private DnsEndPoint? _currentEndPoint;
+
+        public ActiveSubchannelTransport(Subchannel subchannel, TimeSpan socketPingInterval)
```

Comment:
This looks good to me as well, but another set of eyes on the socket handling would be nice.